File: commands/emote.py
#!/usr/bin/env python3
import os
import requests


async def emote(ctx, search, add):

    await ctx.respond(content="⏳ Loading...")

    query = {
        "operationName": "EmoteSearch",
        "query": "query EmoteSearch($query: String, $tags: [String!]!, $sortBy: SortBy!, $filters: Filters, $page: Int, $perPage: Int!, $isDefaultSetSet: Boolean!, $defaultSetId: Id!) {\n  emotes {\n    search(\n      query: $query\n      tags: {tags: $tags, match: ANY}\n      sort: {sortBy: $sortBy, order: DESCENDING}\n      filters: $filters\n      page: $page\n      perPage: $perPage\n    ) {\n      items {\n        id\n        defaultName\n        owner {\n          mainConnection {\n            platformDisplayName\n            __typename\n          }\n          style {\n            activePaint {\n              id\n              name\n              data {\n                layers {\n                  id\n                  ty {\n                    __typename\n                    ... on PaintLayerTypeSingleColor {\n                      color {\n                        hex\n                        __typename\n                      }\n                      __typename\n                    }\n                    ... on PaintLayerTypeLinearGradient {\n                      angle\n                      repeating\n                      stops {\n                        at\n                        color {\n                          hex\n                          __typename\n                        }\n                        __typename\n                      }\n                      __typename\n                    }\n                    ... on PaintLayerTypeRadialGradient {\n                      repeating\n                      stops {\n                        at\n                        color {\n                          hex\n                          __typename\n                        }\n                        __typename\n                      }\n                      shape\n                      __typename\n                    }\n                    ... on PaintLayerTypeImage {\n                      images {\n                        url\n                        mime\n                        size\n                        scale\n                        width\n                        height\n                        frameCount\n                        __typename\n                      }\n                      __typename\n                    }\n                  }\n                  opacity\n                  __typename\n                }\n                shadows {\n                  color {\n                    hex\n                    __typename\n                  }\n                  offsetX\n                  offsetY\n                  blur\n                  __typename\n                }\n                __typename\n              }\n              __typename\n            }\n            __typename\n          }\n          highestRoleColor {\n            hex\n            __typename\n          }\n          __typename\n        }\n        deleted\n        flags {\n          defaultZeroWidth\n          publicListed\n          __typename\n        }\n        imagesPending\n        images {\n          url\n          mime\n          size\n          scale\n          width\n          frameCount\n          __typename\n        }\n        ranking(ranking: TRENDING_WEEKLY)\n        inEmoteSets(emoteSetIds: [$defaultSetId]) @include(if: $isDefaultSetSet) {\n          emoteSetId\n          emote {\n            id\n            alias\n            __typename\n          }\n          __typename\n        }\n        __typename\n      }\n      totalCount\n      pageCount\n      __typename\n    }\n    __typename\n  }\n}",
        "variables": {
            "defaultSetId": "",
            "filters": {"exactMatch": True},
            "isDefaultSetSet": False,
            "page": 1,
            "perPage": 1,
            "query": search,
            "sortBy": "TOP_ALL_TIME",
            "tags": [],
        },
    }

    headers = {
        "authorization": str(os.getenv("7TV_TOKEN")),
        "content-type": "application/json",
    }
    try:
        r = requests.post("https://7tv.io/v4/gql", headers=headers, json=query).json()
        images = r["data"]["emotes"]["search"]["items"][0]["images"]
        webp_2x = next(
            img for img in images if img["mime"] == "image/webp" and img["scale"] == 2
        )
        url = webp_2x["url"]
        name = r["data"]["emotes"]["search"]["items"][0]["defaultName"]
    except:
        content = "❌   **Emote Not Found! Try Again**"
        await ctx.edit(content=content)
        return

    imageReq = requests.get(url)
    image = imageReq.content
    content = url

    if add:
        emote = await ctx.guild.create_custom_emoji(name=name, image=image)
        content = f"✅   **Emote Added To Server**   {emote}"

    await ctx.edit(content=content)

File: commands/emote.ts
import {
  DiscordInteraction,
  InteractionResponse,
  Colors,
} from "../types/discord";
import { embedResponse, errorResponse } from "../utils/discord";
import { getOptionValue } from "../handlers/commands";

interface SevenTVEmote {
  id: string;
  name: string;
  host: {
    url: string;
    files: { name: string; format: string }[];
  };
}

interface SevenTVSearchResponse {
  emotes: SevenTVEmote[];
}

export async function handleEmote(
  interaction: DiscordInteraction,
): Promise<InteractionResponse> {
  const search = getOptionValue<string>(interaction, "search")!;

  try {
    // Search 7TV API
    const url = `https://7tv.io/v3/gql`;
    const query = {
      query: `
        query SearchEmotes($query: String!) {
          emotes(query: $query, limit: 1) {
            items {
              id
              name
              host {
                url
                files {
                  name
                  format
                }
              }
            }
          }
        }
      `,
      variables: { query: search },
    };

    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(query),
    });

    const data = (await response.json()) as {
      data: { emotes: { items: SevenTVEmote[] } };
    };
    const emotes = data.data?.emotes?.items;

    if (!emotes || emotes.length === 0) {
      return errorResponse(`No emotes found for "${search}".`);
    }

    const emote = emotes[0];

    // Get the best quality file (prefer WEBP or GIF)
    const webpFile = emote.host.files.find((f) => f.format === "WEBP");
    const gifFile = emote.host.files.find((f) => f.format === "GIF");
    const file = webpFile || gifFile || emote.host.files[0];

    const emoteUrl = `https:${emote.host.url}/${file?.name || "4x.webp"}`;

    return embedResponse({
      title: "🎭 7TV Emote Search",
      description: `Found emote: **${emote.name}**`,
      color: Colors.PURPLE,
      thumbnail: { url: emoteUrl },
    });
  } catch (error) {
    console.error("7TV API error:", error);
    return errorResponse(
      "Failed to search for emotes. Please try again later.",
    );
  }
}
